client stops working after failing to send messages that are too large

## Requirement - what kind of business use case are you trying to solve?

Sending Jaeger spans.

## Problem - what in Jaeger blocks you from solving the requirement?

After a few unsuccessful sends the client stops sending *any* spans. I've got these error messages by logging the exception as explained in #653 .

There's a couple of these

```
io.jaegertracing.internal.exceptions.SenderException: ThriftSender received a span that was too large, size = 109837, max = 64967
        at io.jaegertracing.thrift.internal.senders.ThriftSender.append(ThriftSender.java:58) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at com.gratex.bl.upm2.config.JaegerRemoteReporter$AppendCommand.execute(JaegerRemoteReporter.java:139) ~[classes/:1.0.1]
        at com.gratex.bl.upm2.config.JaegerRemoteReporter$QueueProcessor.run(JaegerRemoteReporter.java:174) [classes/:1.0.1]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
```

*NOTE*: JaegerRemoteReporter is RemoteReporter + the 1 line for the log.error

After a while I get errors like this:

```
io.jaegertracing.internal.exceptions.SenderException: Failed to flush spans.
        at io.jaegertracing.thrift.internal.senders.ThriftSender.append(ThriftSender.java:76) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at com.gratex.bl.upm2.config.JaegerRemoteReporter$AppendCommand.execute(JaegerRemoteReporter.java:139) ~[classes/:1.0.1]
        at com.gratex.bl.upm2.config.JaegerRemoteReporter$QueueProcessor.run(JaegerRemoteReporter.java:174) [classes/:1.0.1]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
Caused by: io.jaegertracing.internal.exceptions.SenderException: Could not send 1 spans
        at io.jaegertracing.thrift.internal.senders.UdpSender.send(UdpSender.java:66) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.flush(ThriftSender.java:110) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.append(ThriftSender.java:73) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        ... 3 more
Caused by: org.apache.thrift.transport.TTransportException: Message size too large: 1 > 65000
        at io.jaegertracing.thrift.internal.reporters.protocols.ThriftUdpTransport.write(ThriftUdpTransport.java:132) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at org.apache.thrift.protocol.TCompactProtocol.writeByteDirect(TCompactProtocol.java:486) ~[libthrift-0.12.0.jar:0.12.0]
        at org.apache.thrift.protocol.TCompactProtocol.writeByteDirect(TCompactProtocol.java:493) ~[libthrift-0.12.0.jar:0.12.0]
        at org.apache.thrift.protocol.TCompactProtocol.writeFieldBeginInternal(TCompactProtocol.java:260) ~[libthrift-0.12.0.jar:0.12.0]
        at org.apache.thrift.protocol.TCompactProtocol.writeFieldBegin(TCompactProtocol.java:242) ~[libthrift-0.12.0.jar:0.12.0]
        at io.jaegertracing.thriftjava.Tag$TagStandardScheme.write(Tag.java:920) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Tag$TagStandardScheme.write(Tag.java:809) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Tag.write(Tag.java:709) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Span$SpanStandardScheme.write(Span.java:1360) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Span$SpanStandardScheme.write(Span.java:1144) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Span.write(Span.java:1028) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Batch$BatchStandardScheme.write(Batch.java:473) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Batch$BatchStandardScheme.write(Batch.java:407) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thriftjava.Batch.write(Batch.java:344) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.agent.thrift.Agent$emitBatch_args$emitBatch_argsStandardScheme.write(Agent.java:1056) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.agent.thrift.Agent$emitBatch_args$emitBatch_argsStandardScheme.write(Agent.java:1018) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.agent.thrift.Agent$emitBatch_args.write(Agent.java:969) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:71) ~[libthrift-0.12.0.jar:0.12.0]
        at org.apache.thrift.TServiceClient.sendBaseOneway(TServiceClient.java:66) ~[libthrift-0.12.0.jar:0.12.0]
        at io.jaegertracing.agent.thrift.Agent$Client.send_emitBatch(Agent.java:70) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.agent.thrift.Agent$Client.emitBatch(Agent.java:63) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.UdpSender.send(UdpSender.java:64) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.flush(ThriftSender.java:110) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.append(ThriftSender.java:73) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        ... 3 more
```

```
io.jaegertracing.internal.exceptions.SenderException: Failed to flush spans.
        at io.jaegertracing.thrift.internal.senders.ThriftSender.append(ThriftSender.java:76) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at com.gratex.bl.upm2.config.JaegerRemoteReporter$AppendCommand.execute(JaegerRemoteReporter.java:139) ~[classes/:1.0.1]
        at com.gratex.bl.upm2.config.JaegerRemoteReporter$QueueProcessor.run(JaegerRemoteReporter.java:174) [classes/:1.0.1]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
Caused by: io.jaegertracing.internal.exceptions.SenderException: Could not send 8 spans
        at io.jaegertracing.thrift.internal.senders.UdpSender.send(UdpSender.java:66) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.flush(ThriftSender.java:110) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.append(ThriftSender.java:73) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        ... 3 more
Caused by: org.apache.thrift.transport.TTransportException: Message size too large: 1 > 65000
        at io.jaegertracing.thrift.internal.reporters.protocols.ThriftUdpTransport.write(ThriftUdpTransport.java:132) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at org.apache.thrift.protocol.TCompactProtocol.writeByteDirect(TCompactProtocol.java:486) ~[libthrift-0.12.0.jar:0.12.0]
        at org.apache.thrift.protocol.TCompactProtocol.writeMessageBegin(TCompactProtocol.java:206) ~[libthrift-0.12.0.jar:0.12.0]
        at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:70) ~[libthrift-0.12.0.jar:0.12.0]
        at org.apache.thrift.TServiceClient.sendBaseOneway(TServiceClient.java:66) ~[libthrift-0.12.0.jar:0.12.0]
        at io.jaegertracing.agent.thrift.Agent$Client.send_emitBatch(Agent.java:70) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.agent.thrift.Agent$Client.emitBatch(Agent.java:63) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.UdpSender.send(UdpSender.java:64) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.flush(ThriftSender.java:110) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        at io.jaegertracing.thrift.internal.senders.ThriftSender.append(ThriftSender.java:73) ~[jaeger-thrift-1.0.0.jar:1.0.0]
        ... 3 more
```

## Proposal - what do you suggest to solve the problem or improve the existing situation?

A few failing messages shouldn't cause the whole client to fall down. I suspect some internal byte buffer doesn't get reinitialized correctly.

Also, I'd like to have the message size configurable via an environment variable.

## Any open questions to address

None. Thank you!

--- a/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
+++ b/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
@@ -54,15 +54,15 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
 
     io.jaegertracing.thriftjava.Span thriftSpan = JaegerThriftSpanConverter.convertSpan(span);
     int spanSize = calculateSpanSize(thriftSpan);
-    if ((processBytesSize + spanSize) > getMaxSpanBytes()) {
+    if (spanSize > getMaxSpanBytes()) {
       throw new SenderException(String.format("ThriftSender received a span that was too large, size = %d, max = %d",
-          (processBytesSize + spanSize), getMaxSpanBytes()), null, 1);
+          spanSize, getMaxSpanBytes()), null, 1);
     }
 
     byteBufferSize += spanSize;
-    if (byteBufferSize <= getMaxSpanBytes()) {
+    if (byteBufferSize <= getMaxBatchBytes()) {
       spanBuffer.add(thriftSpan);
-      if (byteBufferSize < getMaxSpanBytes()) {
+      if (byteBufferSize < getMaxBatchBytes()) {
         return 0;
       }
       return flush();
@@ -95,6 +95,10 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
     } catch (Exception e) {
       throw new SenderException("ThriftSender failed writing Span to memory buffer.", e, 1);
     }
+  }
+
+  protected int getMaxSpanBytes() {
+    return getMaxBatchBytes() - processBytesSize;
   }
 
   public abstract void send(Process process, List<io.jaegertracing.thriftjava.Span> spans) throws SenderException;

--- a/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
+++ b/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
@@ -115,7 +115,7 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
       throw new SenderException("Failed to flush spans.", e, n);
     } finally {
       spanBuffer.clear();
-      spanBytesSize = processBytesSize;
+      spanBytesSize = 0;
     }
     return n;
   }

--- a/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
+++ b/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
@@ -29,7 +29,7 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
 
   private Process process;
   private int processBytesSize;
-  private int byteBufferSize;
+  private int spanBytesSize;
 
   @ToString.Exclude private List<io.jaegertracing.thriftjava.Span> spanBuffer;
 
@@ -49,7 +49,6 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
       process = new Process(span.getTracer().getServiceName());
       process.setTags(JaegerThriftSpanConverter.buildTags(span.getTracer().tags()));
       processBytesSize = calculateProcessSize(process);
-      byteBufferSize += processBytesSize;
     }
 
     io.jaegertracing.thriftjava.Span thriftSpan = JaegerThriftSpanConverter.convertSpan(span);
@@ -59,10 +58,10 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
           spanSize, getMaxSpanBytes()), null, 1);
     }
 
-    byteBufferSize += spanSize;
-    if (byteBufferSize <= getMaxBatchBytes()) {
+    spanBytesSize += spanSize;
+    if (spanBytesSize <= getMaxSpanBytes()) {
       spanBuffer.add(thriftSpan);
-      if (byteBufferSize < getMaxBatchBytes()) {
+      if (spanBytesSize < getMaxSpanBytes()) {
         return 0;
       }
       return flush();
@@ -77,7 +76,7 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
     }
 
     spanBuffer.add(thriftSpan);
-    byteBufferSize = processBytesSize + spanSize;
+    spanBytesSize = spanSize;
     return n;
   }
 
@@ -116,7 +115,7 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
       throw new SenderException("Failed to flush spans.", e, n);
     } finally {
       spanBuffer.clear();
-      byteBufferSize = processBytesSize;
+      spanBytesSize = processBytesSize;
     }
     return n;
   }

--- a/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
+++ b/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSender.java
@@ -54,9 +54,9 @@ public abstract class ThriftSender extends ThriftSenderBase implements Sender {
 
     io.jaegertracing.thriftjava.Span thriftSpan = JaegerThriftSpanConverter.convertSpan(span);
     int spanSize = calculateSpanSize(thriftSpan);
-    if (spanSize > getMaxSpanBytes()) {
+    if ((processBytesSize + spanSize) > getMaxSpanBytes()) {
       throw new SenderException(String.format("ThriftSender received a span that was too large, size = %d, max = %d",
-          spanSize, getMaxSpanBytes()), null, 1);
+          (processBytesSize + spanSize), getMaxSpanBytes()), null, 1);
     }
 
     byteBufferSize += spanSize;

--- a/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSenderBase.java
+++ b/jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/ThriftSenderBase.java
@@ -37,7 +37,7 @@ public abstract class ThriftSenderBase {
 
   protected final TProtocolFactory protocolFactory;
   private final TSerializer serializer;
-  private final int maxSpanBytes;
+  private final int maxBatchBytes;
 
   @ToString.Exclude private AutoExpandingBufferWriteTransport memoryTransport;
 
@@ -63,13 +63,13 @@ public abstract class ThriftSenderBase {
       maxPacketSize = ThriftUdpTransport.MAX_PACKET_SIZE;
     }
 
-    maxSpanBytes = maxPacketSize - EMIT_BATCH_OVERHEAD;
+    maxBatchBytes = maxPacketSize - EMIT_BATCH_OVERHEAD;
     memoryTransport = new AutoExpandingBufferWriteTransport(maxPacketSize, 2);
     serializer = new TSerializer(protocolFactory);
   }
 
-  protected int getMaxSpanBytes() {
-    return maxSpanBytes;
+  protected int getMaxBatchBytes() {
+    return maxBatchBytes;
   }
 
   protected byte[] serialize(TBase<?,?> thriftBase) throws Exception {

--- a/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/ThriftSenderTest.java
+++ b/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/ThriftSenderTest.java
@@ -50,6 +50,21 @@ public class ThriftSenderTest {
   }
 
   @Test(expected = SenderException.class)
+  public void appendFail() throws Exception {
+    int size = 0;
+    ThriftSender sender = new ThriftSender(ProtocolType.Compact, 0) {
+      @Override
+      public void send(Process process, List<Span> spans) throws SenderException {
+        throw new SenderException("", null, spans.size());
+      }
+    };
+
+    JaegerTracer tracer = new JaegerTracer.Builder("failure").build();
+    sender.append(tracer.buildSpan("flush-fail").start());
+    sender.flush();
+  }
+
+  @Test(expected = SenderException.class)
   public void flushFail() throws Exception {
     ThriftSender sender = new ThriftSender(ProtocolType.Compact, 0) {
       @Override

--- a/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpSenderTest.java
+++ b/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpSenderTest.java
@@ -84,21 +84,30 @@ public class UdpSenderTest {
     reporter.close();
   }
 
-  @Test(expected = SenderException.class)
-  public void testAppendSpanTooLarge() throws Exception {
+  /**
+   * create a mock span with target size
+   */
+  private JaegerSpan buildSpanWithSize(int targetSpanSize) throws Exception {
     JaegerSpan jaegerSpan = tracer.buildSpan("raza").start();
-    String msg = "";
-    for (int i = 0; i < 10001; i++) {
-      msg += ".";
-      jaegerSpan.log(msg);
+    jaegerSpan.setTag("mock", "");
+
+    // contains mock span tag with empty string value, which is later used to match target span size
+    io.jaegertracing.thriftjava.Span emptySpan = JaegerThriftSpanConverter.convertSpan(jaegerSpan);
+
+    int emptySpanSize = sender.getSize(emptySpan);
+    int freePacketSizeLeft = targetSpanSize - emptySpanSize;
+
+    // each "0" takes 1 byte in UTF-8
+    String mockStr = String.join("", Collections.nCopies(freePacketSizeLeft, "0"));
+    jaegerSpan.setTag("mock", mockStr);
+
+    // subtract some chars since the actual encoding takes a few extra bytes
+    while (targetSpanSize != sender.getSize(JaegerThriftSpanConverter.convertSpan(jaegerSpan))) {
+      mockStr = mockStr.substring(1);
+      jaegerSpan.setTag("mock", mockStr);
     }
 
-    try {
-      sender.append(jaegerSpan);
-    } catch (SenderException e) {
-      assertEquals(e.getDroppedSpanCount(), 1);
-      throw e;
-    }
+    return jaegerSpan;
   }
 
   @Test
@@ -129,6 +138,39 @@ public class UdpSenderTest {
     int result = sender.append(jaegerSpan);
 
     assertEquals(expectedNumSpans, result);
+  }
+
+  @Test
+  public void testAppendMaxSize() throws Exception {
+    Process process = new Process(tracer.getServiceName())
+            .setTags(JaegerThriftSpanConverter.buildTags(tracer.tags()));
+
+    int processSize = sender.getSize(process);
+
+    JaegerSpan jaegerSpan = buildSpanWithSize(maxPacketSize - UdpSender.EMIT_BATCH_OVERHEAD - processSize);
+
+    sender.append(jaegerSpan);
+
+    // add a span that overflows the limit to hit the last branch
+    int result = sender.append(jaegerSpan);
+
+    assertEquals(1, result);
+  }
+
+  @Test(expected = SenderException.class)
+  public void testAppendSpanTooLarge() throws Exception {
+    Process process = new Process(tracer.getServiceName())
+            .setTags(JaegerThriftSpanConverter.buildTags(tracer.tags()));
+    int processSize = sender.getSize(process);
+
+    JaegerSpan jaegerSpan = buildSpanWithSize(maxPacketSize - UdpSender.EMIT_BATCH_OVERHEAD - processSize + 1);
+
+    try {
+      sender.append(jaegerSpan);
+    } catch (SenderException e) {
+      assertEquals(e.getDroppedSpanCount(), 1);
+      throw e;
+    }
   }
 
   @Test

--- a/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpSenderTest.java
+++ b/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpSenderTest.java
@@ -149,9 +149,6 @@ public class UdpSenderTest {
 
     JaegerSpan jaegerSpan = buildSpanWithSize(maxPacketSize - UdpSender.EMIT_BATCH_OVERHEAD - processSize);
 
-    sender.append(jaegerSpan);
-
-    // add a span that overflows the limit to hit the last branch
     int result = sender.append(jaegerSpan);
 
     assertEquals(1, result);

--- a/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpSenderTest.java
+++ b/jaeger-thrift/src/test/java/io/jaegertracing/thrift/internal/senders/UdpSenderTest.java
@@ -152,6 +152,11 @@ public class UdpSenderTest {
     int result = sender.append(jaegerSpan);
 
     assertEquals(1, result);
+
+    // test if the buffer is reinitialized correctly
+    result = sender.append(jaegerSpan);
+
+    assertEquals(1, result);
   }
 
   @Test(expected = SenderException.class)

